parser/prism/Translator.cc

```diff
@@ -1938,6 +1919,113 @@ void Translator::patternTranslateMultiInto(NodeVec &outSorbetNodes, absl::Span<p
+
+    parser::NodeVec params;
+
+    auto restSize = paramsNode->rest == nullptr ? 0 : 1;
+    auto kwrestSize = paramsNode->keyword_rest == nullptr ? 0 : 1;
+    auto blockSize = paramsNode->block == nullptr ? 0 : 1;
+
+    params.reserve(requireds.size() + optionals.size() + restSize + posts.size() + keywords.size() + kwrestSize +
+                   blockSize);
+
+    translateMultiInto(params, requireds);
+    translateMultiInto(params, optionals);
+
+    if (paramsNode->rest != nullptr)
+        params.emplace_back(translate(paramsNode->rest));
+
+    translateMultiInto(params, posts);
+    translateMultiInto(params, keywords);
+
+    if (auto prismKwRestNode = paramsNode->keyword_rest; prismKwRestNode != nullptr) {
+        switch (PM_NODE_TYPE(prismKwRestNode)) {
+            case PM_KEYWORD_REST_PARAMETER_NODE: // `**kwargs`
+                [[fallthrough]];
+            case PM_FORWARDING_PARAMETER_NODE: { // `**`
+                params.emplace_back(translate(prismKwRestNode));
+                break;
+            }
+            case PM_NO_KEYWORDS_PARAMETER_NODE: { // `**nil`
+                params.emplace_back(make_unique<parser::Kwnilarg>(translateLoc(prismKwRestNode->location)));
+                break;
+            }
+            default:
+                unreachable("Unexpected keyword_rest node type in Hash pattern.");
+        }
+    }
+
+    if (paramsNode->block != nullptr)
+        params.emplace_back(translate(up_cast(paramsNode->block)));
+
+    return make_unique<parser::Args>(location, move(params));
+}
+
+tuple<ast::MethodDef::ARGS_store, ast::InsSeq::STATS_store, bool /* didDesugarParams */>
+Translator::desugarParametersNode(unique_ptr<parser::Args> &paramsNode, bool attemptToDesugarParams) {
+    if (paramsNode == nullptr) {
+        return make_tuple(ast::MethodDef::ARGS_store{}, ast::InsSeq::STATS_store{}, true);
+    }
+
+    if (!attemptToDesugarParams) {
+        return make_tuple(ast::MethodDef::ARGS_store{}, ast::InsSeq::STATS_store{}, false);
+    }
+
+    auto supportedParams = absl::c_all_of(paramsNode->args, [](auto &param) {
+        // These other block types don't have their own dedicated desugared
+        // representation, so they won't be directly translated. Instead, they have
+        // special desugar logic below.
+        return param->hasDesugaredExpr() ||
+               //   parser::NodeWithExpr::isa_node<parser::Blockarg>(param.get()) ||
+               parser::NodeWithExpr::isa_node<parser::Kwrestarg>(param.get()) ||
+               parser::NodeWithExpr::isa_node<parser::Kwnilarg>(param.get()) ||
+               parser::NodeWithExpr::isa_node<parser::ForwardArg>(param.get());
+    });
+
+    if (!supportedParams) {
+        return make_tuple(ast::MethodDef::ARGS_store{}, ast::InsSeq::STATS_store{}, false);
+    }
+
+    ast::MethodDef::ARGS_store argsStore;
+    ast::InsSeq::STATS_store statsStore;
+
+    for (auto &param : paramsNode->args) {
```

Comment:
This is the logic from `desugarArgs()` in `Desugar.cc`

https://github.com/Shopify/sorbet/blob/bdc6ad5c34271b02fb2633e94ea82d2eb9fe5a79/ast/desugar/PrismDesugar.cc#L59-L103